File: array-string/rotate-array/rotate-array.cs
public class Solution
{
    public void Rotate(int[] nums, int k)
    {
        int n = nums.Length;
        if (n <= 1) return;
        k %= n;
        if (k == 0) return;

        Reverse(nums, 0, n - 1);
        Reverse(nums, 0, k - 1);
        Reverse(nums, k, n - 1);
    }

    private void Reverse(int[] a, int l, int r)
    {
        while (l < r)
        {
            (a[l], a[r]) = (a[r], a[l]);
            l++; r--;
        }
    }

    public void RotateExtraArray(int[] nums, int k)
    {
        int n = nums.Length;
        if (n <= 1) return;
        k %= n;
        if (k == 0) return;

        var tmp = new int[n];
        for (int i = 0; i < n; i++)
        {
            tmp[(i + k) % n] = nums[i];
        }
        Array.Copy(tmp, nums, n);
    }
    
    public void RotateCycle(int[] nums, int k) {
        int n = nums.Length;
        if (n <= 1) return;
        k %= n;
        if (k == 0) return;

        int moved = 0;
        for (int start = 0; moved < n; start++) {
            int curr = start;
            int prev = nums[start];
            do {
                int next = (curr + k) % n;
                (nums[next], prev) = (prev, nums[next]);
                curr = next;
                moved++;
            } while (curr != start);
        }
    }
}

File: array-string/rotate-array/rotate-array.go
func rotate(nums []int, k int)  {
    n := len(nums)
	if n <= 1 {
		return
	}
	k %= n
	if k == 0 {
		return
	}

	reverse(nums, 0, n-1)
	reverse(nums, 0, k-1)
	reverse(nums, k, n-1)
}

func reverse(a []int, l, r int) {
	for l < r {
		a[l], a[r] = a[r], a[l]
		l++
		r--
	}
}

func rotateExtraArray(nums []int, k int) {
	n := len(nums)
	if n <= 1 {
		return
	}
	k %= n
	if k == 0 {
		return
	}

	tmp := make([]int, n)
	for i := 0; i < n; i++ {
		tmp[(i+k)%n] = nums[i]
	}
	copy(nums, tmp)
}

func rotateCycles(nums []int, k int) {
	n := len(nums)
	if n <= 1 {
		return
	}
	k %= n
	if k == 0 {
		return
	}

	moved := 0
	for start := 0; moved < n; start++ {
		curr := start
		prev := nums[start]
		for {
			next := (curr + k) % n
			nums[next], prev = prev, nums[next]
			curr = next
			moved++
			if curr == start {
				break
			}
		}
	}
}
